### PR TITLE
[Feature] Support AMS(Array/Map/Struct) in Hudi MOR on C++ side

### PR DESCRIPTION
**Pull request by dirtysalt**
Signed-off-by: yanz <dirtysalt1987@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiSliceScanner.java [hunk 1]
```diff
@@ -99,6 +99,7 @@ private JobConf makeJobConf(Properties properties) {
         Configuration conf = new Configuration();
         JobConf jobConf = new JobConf(conf);
         jobConf.setBoolean("hive.io.file.read.all.columns", false);
+        jobConf.setBoolean("dfs.client.use.legacy.blockreader", false);
```
Comment (dirtysalt) [inline java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiSliceScanner.java:102]: Without this setting, we get error like

```
java.lang.UnsupportedOperationException: readDirect unsupported in RemoteBlockReader
        at org.apache.hadoop.hdfs.RemoteBlockReader.read(RemoteBlockReader.java:492)
        at org.apache.hadoop.hdfs.DFSInputStream$ByteBufferStrategy.doRead(DFSInputStream.java:789)
        at org.apache.hadoop.hdfs.DFSInputStream.readBuffer(DFSInputStream.java:823)
        at org.apache.hadoop.hdfs.DFSInputStream.readWithStrategy(DFSInputStream.java:883)
        at org.apache.hadoop.hdfs.DFSInputStream.read(DFSInputStream.java:938)
        at org.apache.hadoop.fs.FSDataInputStream.read(FSDataInputStream.java:143)
        at org.apache.parquet.hadoop.util.H2SeekableInputStream$H2Reader.read(H2SeekableInputStream.java:81)
        at org.apache.parquet.hadoop.util.H2SeekableInputStream.readFully(H2SeekableInputStream.java:90)
        at org.apache.parquet.hadoop.util.H2SeekableInputStream.readFully(H2SeekableInputStream.java:75)
        at org.apache.parquet.hadoop.ParquetFileReader.readFooter(ParquetFileReader.java:575)
        at org.apache.parquet.hadoop.ParquetFileReader.readFooter(ParquetFileReader.java:527)
        at org.apache.parquet.hadoop.ParquetFileReader.readFooter(ParquetFileReader.java:521)
        at org.apache.parquet.hadoop.ParquetFileReader.readFooter(ParquetFileReader.java:469)
        at org.apache.parquet.hadoop.ParquetFileReader.readFooter(ParquetFileReader.java:454)
        at org.apache.hadoop.hive.ql.io.parquet.ParquetRecordReaderBase.getSplit(ParquetRecordReaderBase.java:79)
        at org.apache.hadoop.hive.ql.io.parquet.read.ParquetRecordReaderWrapper.<init>(ParquetRecordReaderWrapper.java:75)
        at org.apache.hadoop.hive.ql.io.parquet.read.ParquetRecordReaderWrapper.<init>(ParquetRecordReaderWrapper.java:60)
        at org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat.getRecordReader(MapredParquetInputFormat.java:92)
        at org.apache.hudi.hadoop.HoodieParquetInputFormat.getRecordReaderInternal(HoodieParquetInputFormat.java:89)
        at org.apache.hudi.hadoop.HoodieParquetInputFormat.getRecordReader(HoodieParquetInputFormat.java:83)
        at org.apache.hudi.hadoop.realtime.HoodieParquetRealtimeInputFormat.getRecordReader(HoodieParquetRealtimeInputFormat.java:74)
        at com.starrocks.hudi.reader.HudiSliceScanner.initReader(HudiSliceScanner.java:177)
        at com.starrocks.hudi.reader.HudiSliceScanner.open(HudiSliceScanner.java:195)
```

Comment (dirtysalt) [reply java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiSliceScanner.java:102]: It's very wierd that after re-compilation, if I remove this line, everything works fine.

**Review by Smith-Cruise — COMMENTED** [review]
You can add some BE UT about partial load for complex types.

For example:
1: Test different order in struct subfield. Like TypeDescriptor struct<a int, b int>, data type in hudi struct<b int, a int>
2: Test case-sensitive about struct subfield. (Struct subfield name is case-sensitive)
3: Partial load in struct & map type

**Review by Smith-Cruise — DISMISSED** [review]

File: java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java [hunk 2]
```diff
@@ -134,12 +135,22 @@ public void unpackMap(List<ColumnValue> keys, List<ColumnValue> values) {
...
+            }
+        }
         for (int i = 0; i < structFieldIndex.size(); i++) {
-            StructField sf = fields.get(structFieldIndex.get(i));
-            Object o = inspector.getStructFieldData(fieldData, sf);
+            Integer idx = structFieldIndex.get(i);
             HudiColumnValue cv = null;
-            if (o != null) {
-                cv = new HudiColumnValue(sf.getFieldObjectInspector(), o);
+            if (idx != null) {
```
Comment (miomiocat) [inline java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java:148]: when is `idx` NULL

File: java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java [hunk 3]
```diff
@@ -134,12 +135,22 @@ public void unpackMap(List<ColumnValue> keys, List<ColumnValue> values) {
     public void unpackStruct(List<Integer> structFieldIndex, List<ColumnValue> values) {
         StructObjectInspector inspector = (StructObjectInspector) fieldInspector;
         List<? extends StructField> fields = inspector.getAllStructFieldRefs();
+        // if we don't do any prune, we preserve all fields.
+        if (structFieldIndex == null) {
+            structFieldIndex = new ArrayList<>(fields.size());
+            for (int i = 0; i < fields.size(); i++) {
+                structFieldIndex.add(i);
+            }
+        }
```
Comment (miomiocat) [inline java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java:144]: place this logic to `pruneOnStructSelectedFields`

File: java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java (hunk 2 above)
Comment (dirtysalt) [reply java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java:148]: When field is not found. For example,  BE thinks struct is `<B:int, a:int>`, but actually in hudi it's `<b:int, a:int>`'

File: java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java (hunk 3 above)
Comment (dirtysalt) [reply java-extensions/hudi-reader/src/main/java/com/starrocks/hudi/reader/HudiColumnValue.java]: Updated.

**Review by miomiocat — DISMISSED** [review]
LGTM

**Review by Smith-Cruise — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by Youngwb — APPROVED** [review]

**Review by zombee0 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by stephen-shelby — APPROVED** [review]

**Comment by dirtysalt** [comment]
run starrocks_be_unittest

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Pull request merged**